quinn-proto/src/connection/mod.rs

```diff
@@ -643,97 +640,73 @@ impl Connection {
+                            || buf.datagram_start_offset() + buf.segment_size()
+                                > buf.datagram_max_offset()
                         {
                             trace!(
                                 "GSO truncated by demand for {} padding bytes or loss probe",
-                                segment_size - packet_len_unpadded
+                                buf.segment_size() - packet_len_unpadded
                             );
                             builder_storage = Some(builder);
                             break;
                         }
 
                         // Pad the current datagram to GSO segment size so it can be included in the
                         // GSO batch.
-                        builder.pad_to(segment_size as u16);
+                        builder.pad_to(buf.segment_size() as u16);
                     }
 
-                    builder.finish_and_track(now, self, sent_frames.take(), buf);
-
-                    if num_datagrams == 1 {
-                        // Set the segment size for this GSO batch to the size of the first UDP
-                        // datagram in the batch. Larger data that cannot be fragmented
-                        // (e.g. application datagrams) will be included in a future batch. When
-                        // sending large enough volumes of data for GSO to be useful, we expect
-                        // packet sizes to usually be consistent, e.g. populated by max-size STREAM
-                        // frames or uniformly sized datagrams.
-                        segment_size = buf.len();
-                        // Clip the unused capacity out of the buffer so future packets don't
-                        // overrun
-                        buf_capacity = buf.len();
-
-                        // Check whether the data we planned to send will fit in the reduced segment
-                        // size. If not, bail out and leave it for the next GSO batch so we don't
-                        // end up trying to send an empty packet. We can't easily compute the right
-                        // segment size before the original call to `space_can_send`, because at
-                        // that time we haven't determined whether we're going to coalesce with the
-                        // first datagram or potentially pad it to `MIN_INITIAL_SIZE`.
-                        if space_id == SpaceId::Data {
-                            let frame_space_1rtt =
-                                segment_size.saturating_sub(self.predict_1rtt_overhead(Some(pn)));
-                            if self.space_can_send(space_id, frame_space_1rtt).is_empty() {
-                                break;
-                            }
+                    builder.finish_and_track(now, self, sent_frames.take(), &mut buf);
+
+                    if buf.num_datagrams() == 1 && space_id == SpaceId::Data {
```

Comment:
This if condition shouldn't have been collapsed.  The buffer needs to be clipped in all packet spaces.  The extra size check is only needed in the data packet space.